## Abstract {.page_break_before}

The current reference genome is the backbone of diverse and rich annotations. 
Simple text formats, like VCF or BED, have been widely adopted and helped this critical exchange of genomic information. 
To enable similar enrichment of a pangenome reference, there is a dire need for tools and formats for pangenomic annotation. 
The Graph Alignment Format (GAF) is a text format, tab-delimitted like BED/VCF files, which was proposed to represent alignments
While tt could be used to represent any type of annotation in a pangenome graph, there is no tools to query them efficiently.

Here, we present extension to vg and HTSlib that provide efficient sorting, indexing and querying for GAF files.
With this approach, annotations overlapping a subgraph can be extracted quickly.
Paths are sorted based on the IDs of traversed nodes, compressed with BGZIP, and indexed with HTSlib/tabix that we extended to work with the GAF format. 
In addition, we updated `vg annotate` to better produce graph annotations as paths, starting from annotation files relative to linear references. 
More precisely, it can to take annotations in BED or GFF3 files, written relative to reference paths or haplotypes, and produce GAF files representing their paths through the pangenome.

To showcase these commands, we projected annotations for all haplotypes in the latest draft human pangenome (HPRC v1.1 GRCh38-based Minigraph-Cactus pangenome). 
This included genes, segmental duplications, tandem repeats and repeats annotations. 
`vg annotate` can annotate ~4M gene annotations in ~16 mins, and ~5.5M repeats from RepeatMasker in ~9 mins on a single-threaded machine. 
Finally, these rich annotations can then be quickly queried with `vg` and visualized using existing tools like the sequenceTubeMap or Bandage.


