<link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.7.2/css/all.css">

## Introduction

<!-- Non-exhaustive tool list at https://pangenome.github.io/ -->

The current reference genome is the backbone of diverse and rich annotations. 
It is used as a reference to map sequencing reads.
The coordinate system that it provides is used by large annotation databases gathering functional elements, known variant information, genomic elements. 
Over the past decades, the organization and visualization of these annotations has been central for understanding and sharing results from genomic studies.
In practice, annotations are typically saved in additional files following a text format that can be compressed and indexed for fast query (e.g. VCF, BED, GFF).
Under the hood, the HTSlib library supports the indexing of the most used file formats[@htslib].
These annotation files are easy to write and to load in software like IGV[@igv] or website like the UCSC Genome Browser[@doi:10.1093/nar/gkac1072].

With the improved genome sequencing technologies, more high-quality genomes can be produced and combined into pangenomes.
A pangenome represents multiple genomes, often as a graph describing adjancency (edges) of the sequences (nodes) in the genomes.
Tools to work with such pangenomes are still in their infancy although there are now several options that shows improved performance over traditional approaches.
In particular, sequencing data analysis benefits from using pangenomes as a reference for read mapping[@vg;@giraffe], variant calling[@vgsv], peak calling[@doi:10.1371/journal.pcbi.1006731;@doi:10.1186/s13059-020-02038-8], or transcript expression quantification[@doi:10.1038/s41592-022-01731-9].
This means that we are now manipulating genomic objects like sequencing reads, epigenetic regions, genomic variants, in the pangenome graph space.
Currently, those results are typically projected to the linear reference genome.
As for linear genome reference, it will be essential to be able to organize and visualize genomic annotation in the pangenome.
A lack of user-friendly querying and visualization options would hamper the adoption of the pangenomic paradigm.

Several interactive visualization tools for pangenomes already exist but mostly focus on representing the graph topology with specialized integration of additional information layers. 
Bandage is an interactive assembly graph visualization tool which can scale to pangenomes up to hundreds of thousands of nodes on modern computers[@bandage].
GfaViz is another interactive sequence graph visualization tool which supports the GFA 2 format and its new features[@doi:10.1093/bioinformatics/bty1046].
In particular, it can represent the different paths annotated in the GFA file.
GfaViz cannot load additional annotation files: the annotations, written for example as GFA *paths*, must be included in the main pangenome file to be accessible during exploration. 
<!-- [`gfaestus`](https://github.com/chfi/gfaestus)???. -->
The sequenceTubeMap visualizes the pangenome and sequencing reads aligned to it, allowing the user to query specific regions[@tubemap].
MoMI-G, or Modular Multi-scale Integrated Genome Graph Browser, focuses on the visualization of structural variants[@doi:10.1186/s12859-019-3145-2].
Variants are interactively selected and represented in a pangenome view a custom sequenceTubeMap representation. 
Supporting reads and genome annotations can also be included in that representation. 
Currently, only the reference genome path can be annotated from files in BED/GFF/WIG format.
Panache is another pangenomic tool specializing in gene-centric visualization of a linearized view of the pangenome, where blocks of homologous sequences are represented side by side[@doi:10.1093/bioinformatics/btab688].
The blocks are interactively explored in a web-based application to explore their gene content and their absence/presence in the different genimes.
Panache is notably used in the Banana Genome Hub[@doi:10.1093/database/bat035].
In summary, some visualization tools exist but it is not yet clear how to best provide additional annotation layers to their graph representation.

A few options also exist to display static representations of a pangenome graph (or subgraph).
The `vg` toolkit can output a pangenome and alignments in the DOT format. 
Paths can be displayed too, although the image can become hard to read when many paths are included or in large grahs.
The `odgi` toolkit offers visualizations that scale better to larger graphs and annotations[@odgi]. 
The pangenome is, for example, linearized using an 1D layout[@layout] and annotations displayed on top of the graph as an array/heatmap.
Of note, `odgi` implemented two options to add annotations from external BED files: one to convert annotations to a CSV file to help color nodes in Bandage, another by injecting new paths into a graph before visualization.
One limitation to injecting new paths into a pangenome is that paths must then start at the beginning of a node as we cannot provide an *offset* value to specify the base at which a path must start.

There is a dire need for a format supporting annotations in the pangenome and that can is easy to create, index and query.
As demonstrated by the success of the BED, GFF or VCF formats, it could help the critical exchange of (pan)genomic information, and allow pangenomic tools access additional information in the pangenomic space.
The Graph Alignment Format (GAF) text format, which was proposed to represent alignments, could be used to represent any type of annotation in a pangenome graph.
It's the lack of techniques to compress, index and query it that limits its adoption at a larger scale.
Here, we present new features of the `vg` and HTSlib ecosystem that provide efficient indexing and querying for pangenomic annotations represented as paths in the GAF format.
We illustrate its values in several applications: projecting and visualizing gene and repeat annotations, summarizing open-chromatin from epigenomics sequencing datasets, or positioning known variants in the pangenome.
