<link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.7.2/css/all.css">

## Introduction

The current reference genome is the backbone of diverse and rich annotations. 
It is used as a reference to map sequencing reads.
The coordinate system that it provides is used by large annotation databases gathering functional elements, known variant information, genomic elements. 
Over the past decades, the organization and visualization of these annotations has been central for understanding and sharing results from genomic studies.

With the improved genome sequencing technologies, more high-quality genomes can be produced and compiled into pangenomes.
A pangenome represents multiple genomes, often as a graph describing adjancency (edges) of the sequences (nodes) in the genomes.
Tools to work with such pangenomes are still in their infancy although there are now several options that shows improved performance over traditional approaches.
In particular, sequencing data analysis benefits from using pangenomes as a reference for read mapping[@vg;@giraffe], variant calling[@vgsv], peak calling[@doi:10.1371/journal.pcbi.1006731;@doi:10.1186/s13059-020-02038-8], or transcript expression quantification[@doi:10.1038/s41592-022-01731-9].
This means that we are now manipulating genomic objects like sequencing reads, epigenetic regions, genomic variants, in the pangenome graph space.
Currently, those results are typically projected to the linear reference genome.
As for linear genome reference, it will be essential to be able to organize and visualize genomic annotation in the pangenome.
A lack of user-friendly querying and visualization options would hamper the adoption of the pangenomic paradigm.

Several visualization tools for pangenomes already exist but mostly focus on representing the graph topology with limited or specialized integration of additional layers of information. 
The sequenceTubeMap visualizes the pangenome and sequencing reads aligned to it, allowing the user to query specific regions[@tubemap].
Bandage[@bandage] is an interactive assembly graph visualization tool which can scale to pangenomes up to `?`{.red} nodes.
`review more tools`{.red}

To enable similar enrichment of a pangenome reference, there is a dire need for tools and formats for pangenomic annotation. 
Simple text formats, like VCF or BED, have been widely adopted and helped this critical exchange of genomic information. 
The Graph Alignment Format (GAF) text format, which was proposed to represent alignments, could be used to represent any type of annotation in a pangenome graph.
Here we present new features of the `vg` ecosystem that provide indexing, querying, and visualization capabilities for annotations represented as paths.

GraphAligner, vg giraffe[@giraffe].
