## Results

### Sorting and indexing short sequencing reads

Sort 30x Illumina short-read dataset (~300M paired-end reads) GAM File size: 69G sorted Sorting: ~9h and ~2.5Gb mem, GAF: File size: 24 Gb sorted Sorting: ~6h and ~3.5 Gb mem. GAM to GAF conversion: ~2h.

### Annotation of a human pangenome

Human Pangenome Reference Consotium[@hprc]

To showcase these commands, we projected annotations for all haplotypes in the latest draft human pangenome (HPRC v1.1 GRCh38-based Minigraph-Cactus pangenome). 
This included genes, segmental duplications, tandem repeats and repeats annotations. 
`vg annotate` can annotate ~4M gene annotations in ~16 mins, and ~5.5M repeats from RepeatMasker in ~9 mins on a single-threaded machine. 
Finally, these rich annotations can then be quickly queried with `vg` and visualized using existing tools like the sequenceTubeMap or Bandage.

### Coverage of ?? functional datasets from ENCODE

We aligned ?ATAC-seq? datasets from ?? cell types to the draft human pangenome to produce coverage tracks as indexed GAF files. 
On average, there were about ?? paths representing high read coverage which were ?? nodes (?? bases) long. 
Thousands of paths with high ATAC-seq read coverage traversed more than three nodes, i.e. regions of the pangenome with variation (see @tbl:coverage_summary).

| Dataset         | paths | average bases | average nodes | traversing >2 nodes |
|:----------------|:-----:|---------------|---------------|---------------------|
| Exp1 Cell typeA | ??    |               |               | ?? (??%)            |
| Exp1 Cell typeA | ??    |               |               |                     |
| Exp1 Cell typeA | ??    |               |               |                     |
| Exp1 Cell typeA | ??    |               |               |                     |
| Exp1 Cell typeA | ??    |               |               |                     |

Table: High coverage tracks from ?? functional datasets on the HPRC pangenome. 
{#tbl:coverage_summary}


It took on average ?? cpu.hours to map the reads to the pangenome using VG Giraffe, and ?? cpu.hours to produce the coverage tracks.
Sorting, compressing and indexing them took only ?? cpu.hours, on average.
Table @tbl:coverage_benchmark compiles the runtimes and memory used for each step across all samples.

`example of what we could look for and describe`{.red}
Fig @fig:cov_examples shows examples of those tracks visualized using the sequenceTubeMap.
In Fig ??a, the promoter of the *??* gene is seen to be open in the ?? cell type. 
Thanks to the pangenomic view, we see differential coverage of the functional tracks across the variants in the region.
For instance, we notice a small insertion-deletion (indel) where the alternate allele is only covered by 2 reads, while the reference allele is covered by more than 30 reads.
Fig ??b highlights a structural variant, a ??bp insertion, that is highly covered by ATAC-seq in several cell types.
The RepeatMasker annotation in this region, also extracted from an indexed GAF file, flags this insertion as a ?? transposable element.
?? can indeed attract TF?? that lead to open chromatin ?REF?.

![
**Coverage tracks visualized interactively using the sequenceTubeMap.**
a) Promoter... b) Structural variant....
](figures/wide.png "Wide image"){#fig:cov_examples}


### Examples

Using sequenceTubeMap, haplotypes, read alignments and paths can be visualized interactively. 
Hovering on a path displays its name, here the ID of a coding region of the BOLA2B gene.

> Haplotypes: CHM13 (purple), HG00621 (greys). Annotated CDS for HG00621 hap 1 (reds) and 2 (blues).

Using BandageNG, a fork that can import paths in GAF files, paths were searched and colored to illustrate a mobile element insertion.

> Yellow: AluYa5 element annotated in the haplotype 1 of HG00438. Blue: CHM13 reference path.
