## Results

### Sorting and indexing short sequencing reads

Read sorting was tested on Illumina HiSeq paired-end short reads for the HG002 sample, each 150 bp long and a genome coverage of about 30X.
The gzipped GAF file with about 682 million reads was sorted in 6h32 using a single thread and about 2Gb of memory (Table @tbl:readsorting_summary).
Indexing the sorted GAF with `tabix` took 18 mins.
We compared that approach with the current implementation in `vg` which uses files in the GAM format.
GAM is a protobuf alignemnt format introduced in Garrison et al.[@vg].
Sorting a GAM with the same reads took 11h47 using a single thread and about 6 Gb of memory. 

In addition to being about twice as fast to sort, reads written in the GAF format (and bgzipped) also take about half the disk space (52G vs 108G).
The main reason for this reduced space is that GAF doesn't save the full read sequence, only the path through the pangenome and edits to reconstruct it.
GAM files produced by `vg giraffe` can also save additional information as annotations, like the mapping time of each read, that are currently not kept when converting to the GAF format.
Overall, the bgzipped GAF files are half as small and twice as fast to sort for short sequencing reads.

Once indexed, extracting a slice of the pangenome is as efficient as extracting a slice of an indexed BAM, VCF, or BED file in a genomic regions, as it uses the same approach.
For example, extracting reads for ten thousand random regions in the pangenome took about 0.07 second per region to retrieve an average of 1707 reads.
For comparison, the same extraction took on average 0.8 second per region using the GAM format. 

| Format | Time (H:M:S) | Max. memory used (Kb) | File size (Gb) |
|:------:|-------------:|----------------------:|---------------:|
| GAM    |     11:46:58 |              6,236.60 |            108 |
| GAF    |      6:50:28 |              1,904.83 |             52 |

Table: Resources used to sort short sequencing reads for a 30x coverage Illumina human dataset.
{#tbl:readsorting_summary}

### Annotation of a human pangenome

Human Pangenome Reference Consotium[@hprc]

To showcase these commands, we projected annotations for all haplotypes in the latest draft human pangenome (HPRC v1.1 GRCh38-based Minigraph-Cactus pangenome). 
This included genes, segmental duplications, tandem repeats and repeats annotations. 
`vg annotate` can annotate ~4M gene annotations in ~16 mins, and ~5.5M repeats from RepeatMasker in ~9 mins on a single-threaded machine. 
Finally, these rich annotations can then be quickly queried with `vg` and visualized using existing tools like the sequenceTubeMap or Bandage.
Using BandageNG, a fork that can import paths in GAF files, paths were searched and colored to illustrate a mobile element insertion.

![
**Visualization examples with BandageNG**
Example of a AluYa5 transposon insertion (*red*) within the coding sequence of the *PRAMEF4* gene (*blue*).
Both annotations were originally produced at the haplotype level by the Human Pangenome Reference Consoritium.
We projected them to the pangenome, indexed them, and queried a small region to visualize with BandageNG.
The nodes were colored based on those annotations, loaded as paths by BandageNG.
](figures/PRAMEF4.AluInsertion.png "Bandage example"){#fig:bandage}

We also matched and annotated more than 660 thousand variants from the GWAS catalog[@doi:10.1093/nar/gkac1010] and expression QTLs from the GTEx catalog[@doi:10.1038/ng.2653] across 49 tissues (on average 1.45 million variants per tissue).
On average, 94% variants were found in the HPRC pangenome.
The variants files in GAF take only 907Mb of space and can be queried fast for visualization in the sequenceTubeMap or Bandage.
This annotation is showcased in example ?? described in more detail below (see *Example??* section).
Of note, it is also straightforward to convert genotypes for variants from the pangenome to annotated paths. 
This could be the case when the pangenome is used as the backbone of the genotyping of new samples, for example with Pangenie or *vg call*.
To illustrate this use case, we genotyped HG002 using short Illumina reads aligned on the draft human pangenome with vg giraffe and vg call.
The predicted genotypes were converted to GAF and indexed.
They could be visualized using the sequenceTubeMap, for example, along with the aligned reads. 
This type of annotation can help a developer explore variant calls and aligned reads. 
A example is shown in figure ?? and described in more details below.
On the traditional reference genome, the equivalent would be to load both reads and variants in IGV[@igv].

![
**Visualization examples with the sequenceTubeMap**
**A** Projection of the HPRC gene annotation, here visualizing a coding sequence (CDS) of the *CFD* gene.
There is an insertion (see *insertion node* in the middle), that the reference path (*violet* on top) skips.
Most CDS annotations are similar (*blue* paths), but the CSD from the haplotype with the insertion seems to be split in two parts, highlighted in *green*.
**B.** Projecting known variants in the pangenome. 
*black*: Reference path (GRCh38), *pale colors*: other human haplotypes, *reds*: GWAS catalog, *blues*: eQTLs across tissues (GTEx).
**C.** Aligned reads and genotype calls.
*yellow*/*green*: annotation paths from `vg call` genotypes. *reds*/*blues*: short sequencing reads.
**D.** ATAC-seq coverage tracks for 7 tissues from the ENCODE project. 
The reference path (*violet*) and other HPRC haplotypes (*greys*) are shown on top and traverse two variation sites.
ATAC-seq coverage track for the different tissues are shown in different *colors* with different opacity representing coverage level.
The thyroid track (*blue*) is more opaque than others suggesting that this region is more open in this tissue.
The *red* tracks at the bottom show the position of exons from the *TPO* gene, a thyroid-specific gene.
](figures/tubemap.examples.png "Tubemap examples"){#fig:tubemap}

Using sequenceTubeMap, haplotypes, read alignments and paths can be visualized interactively. 

### Coverage of seven functional datasets from ENCODE

We aligned ATAC-seq datasets from 7 cell types to the draft human pangenome to produce coverage tracks as indexed GAF files. 
On average, there were about 475 thousand paths representing high read coverage which were 2.6 nodes (104.8 bases) long. 
On average, 63 thousand paths with high ATAC-seq read coverage traversed more three or more nodes, i.e. regions of the pangenome with variation (see Table @tbl:coverage_summary).

| Dataset                    |   Paths | Average bases | Average nodes | Traversing >2 nodes |
|:---------------------------|--------:|--------------:|--------------:|--------------------:|
| Breast epithelium          | 570,155 |        109.80 |          2.68 |      65,095 (11.4%) |
| Gastrocnemius medialis     | 342,830 |         94.47 |          2.43 |      53,595 (15.6%) |
| Gastroesophageal sphincter | 555,618 |        115.69 |          2.73 |        72,407 (13%) |
| Peyerâ€™s patch              | 270,094 |         97.88 |          2.62 |      43,591 (16.1%) |
| Sigmoid colon              | 678,380 |        115.10 |          2.69 |      82,906 (12.2%) |
| Spleen                     | 531,910 |        104.97 |          2.66 |      67,153 (12.6%) |
| Thyroid gland              | 377,157 |         95.63 |          2.41 |        56,441 (15%) |

Table: High coverage tracks from seven functional datasets on the HPRC pangenome. For each sample, the table shows how many paths had a mean coverage of at least 10 reads, and how long they were.
{#tbl:coverage_summary}

It took on average 7 cpu.hours to map the reads to the pangenome using VG Giraffe, and 2.8 cpu.hours to produce the coverage tracks.
Sorting, compressing and indexing them took only 0.23 cpu.hours, on average.
Table @tbl:coverage_benchmark compiles the runtimes and memory used for each step across all samples.

`example of what we could look for and describe`{.red}
Fig @fig:cov_examples shows examples of those tracks visualized using the sequenceTubeMap.
In Fig ??a, the promoter of the *??* gene is seen to be open in the ?? cell type. 
Thanks to the pangenomic view, we see differential coverage of the functional tracks across the variants in the region.
For instance, we notice a small insertion-deletion (indel) where the alternate allele is only covered by 2 reads, while the reference allele is covered by more than 30 reads.
Fig ??b highlights a structural variant, a ??bp insertion, that is highly covered by ATAC-seq in several cell types.
The RepeatMasker annotation in this region, also extracted from an indexed GAF file, flags this insertion as a ?? transposable element.
?? can indeed attract TF?? that lead to open chromatin ?REF?.

<!-- ![ -->
<!-- **Coverage tracks visualized interactively using the sequenceTubeMap.** -->
<!-- a) Promoter... b) Structural variant.... -->
<!-- ](figures/wide.png "Wide image"){#fig:cov_examples} -->
