## Discussion

The tools and applications described here present an option to streamline the production of annotations in the GAF format, their indexing and efficient querying.
First, these techniques will help integrate additional information into existing or future pangenomic tools.
They will simplify the integration of information like gene or repeat annotations, known variants, or more generally other functional annotations, in pangenomic analysis.
These analyses will be able to work with those annotations in light of underlying genomic variation recapitulated in the pangenome.
Visualization is a specific example where the users typically want to include several layers of annotations, including some custom-made.
Tools to visualize pangenomes will greatly benefit from a simple pipeline to create or load pangenomic annotations.
Second, this work is critical to allow the community to make, share and reuse annotations in the pangenome space.
The GAF format is already used by multiple independent tools, although mostly read mappers.
Thanks to our work, GAF files can now be indexed and queried efficiently, like BED, VCF, or GFF on a linear reference genome. 
We also showcase how annotations on a linear genome can be converted to a path in the pangenome in the GAF format.
For these reasons, we believe it could become the de facto format to represent annotations in the pangenome and accelerate the adoption of the new pangenomic paradigm by the broader genomics field.
Although informative annotations could be analyzed already, the current approach has some limitations.
<!-- - Simplistic handling of clipped paths. -->

The indexing scheme relies on integer node IDs, compacted relative to their position in the pangenome.
If node IDs are not integer, it adds a layer of complexity for the user as they will need to convert them to integers and keep track of the translation between original pangenome and new pangenome.
Both `vg` and `odgi` offer ways to convert pangenome to a new pangenome with compact integer node IDs.

The current implementation is designed with short paths in mind, where the user wants to extract the full annotated paths in a region.
It is convenient when working with short reads, gene annotations, and most genomic repeats, for example.
Larger genomic regions, like chromosome bands, large assembled contigs, or large segmental duplications, can still be represented and manipulated, but might be less practical to work with.
With the current implementation, the full annotated region will be extracted when querying an overlapping region.
The extracted annotations are not clipped to the specified range which could be inconvenient to a user who zooming into a small region but would still like to know about the much larger annotated path traversing this subgraph.
To address this use case, we plan to implement an extraction mode where output annotations are trimmed to keep only the queried subgraph.

The naive metadata integration is another limitation.
Indeed, the annotation information is currently reduced to a single label.
In our applications, those label would contain for example a gene name and haplotype names, or a read coverage bin and average coverage value, all in one label.
This was easy to implement and sufficient for now because there are no tools that handle more advanced metadata organization.
For many annotations, it would be useful to keep the metadata better organized, so that the user can access/use it within visualization tools.
The different metadata could be saved using optional tags that can be added at the end of each GAF record.

Our approach to convert annotations from a linear genome to the pangenome assumes that we have annotations of the different haplotypes in the pangenome.
There is still no clear solution to lift annotations from one reference/haplotype to other haplotypes in the pangenome, except through reanalysis/reannotation of each haplotype.
The homology information embedded in the pangenome could potentially be used to propagate annotations from one haplotype to others more easily. 
This strategy can already be used by annotation tools like CAT[@doi:10.1101/gr.233460.117], as an additional source of gene annotation evidence.
In the future, these techniques might help propagate other types of annotations across the pangenome more efficiently than by reanalyzing the raw data from scratch on each haplotype.
`something about odgi paths/untangle maybe`{.red}

It also highlights the limitations of the existing tools to integrate these files.
Some tools, like Bandage-NG or GfaViz, require manual pre-processing, for example extracting a subgraph and integrating the annotations as embedded paths.
The Sequence Tube Map can now handle indexed bgzipped GAF files, but the query time for large pangenome remains long in practice.
Defining and integrating annotation metadata into its interface will also require a significant amount of development.
Overall, we stress the need for visualization tools that can efficiently layout and organize many paths through a pangenome.

We showed that the GAF format, thanks to new tools for their efficient manipulation, could offer a path for the future of annotations in pangenome graphs.
While it provides an important building block, it is clear that more needs to be done to make it a useful solution for the community.


## Code and data availability

`vg` is available at [https://github.com/vgteam/vg](https://github.com/vgteam/vg).
Bandage-NG is available at [https://github.com/asl/BandageNG](https://github.com/asl/BandageNG).
The Sequence Tube Map is hosted at [https://github.com/vgteam/sequenceTubemap](https://github.com/vgteam/sequenceTubemap)

The analysis presented in this manuscript is documented in the repository at [https://github.com/jmonlong/manu-vggafannot](https://github.com/jmonlong/manu-vggafannot)[@repo].
It contains scripts used to prepare the annotation files, commands and automated pipelines used to annotate them in the pangenome, and helper scripts to summarize the output files.

The annotations of the HPRC v1.1 pangenome were deposited on Zenodo at `??ZENODO_LINK??`{.red}.
This includes gene annotations, repeats from RepeatMasker, simple repeats and segmental duplications.
A coverage track for the seven ENCODE ATAC-seq samples is also available in this repository.

## Acknowledgments

We would like the thank the ENCODE consortium and the laboratory of Michael Snyder for making the ATAC-seq datasets available.

## Author contribution statement

AMN, GH, JM, TTY, and EG contributed code to the vg toolkit (*annotate*, *pack*, *call* subcommands) with support from BP.
AMN integrated new code to HTSlib.
AMN and DC implemented new features in the Sequence Tube Map.
SD selected and analyzed the ATAC-seq datasets from ENCODE.
JM conceived the study, analyzed data, and drafted the manuscript.
All authors contributed in reviewing the text and content of the manuscript.
